pkg/controllers/deprovisioning/consolidation.go

```diff
@@ -58,106 +71,146 @@ func (c *Consolidation) ShouldDeprovision(_ context.Context, n *state.Node, prov
-		if cmd := c.deleteEmpty(ctx, candidates...); cmd.action == actionDelete {
-			return cmd, nil
-		}
+	cmd, validationTimestamp, err = c.deleteEmpty(ctx, validationTimestamp, candidates...)
+	if err != nil {
+		return Command{action: actionFailed}, fmt.Errorf("deleting empty nodes for consolidation, %w", err)
+	}
+	// If there are empty nodes, delete the empty nodes for this provisioning loop.
+	if cmd.action == actionDelete {
+		return cmd, nil
 	}
 
 	pdbs, err := NewPDBLimits(ctx, c.kubeClient)
 	if err != nil {
 		return Command{action: actionFailed}, fmt.Errorf("tracking PodDisruptionBudgets, %w", err)
 	}
-	// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
-	// scheduling unless its really needed
-	if !canBeTerminated(candidates[attempt], pdbs) {
-		return Command{action: actionNotPossible}, nil
-	}
 
-	return c.computeConsolidation(ctx, candidates[attempt])
+	for _, candidate := range candidates {
+		// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+		// scheduling unless its really needed
+		if !canBeTerminated(candidate, pdbs) {
+			continue
+		}
+		// do a simulated scheduling loop to check if Consolidation is possible
+		cmd, validationTimestamp, err = c.computeConsolidation(ctx, validationTimestamp, candidate)
+		if err != nil {
+			return Command{action: actionFailed}, err
+		}
+
+		if err := c.waitForTTL(ctx, validationTimestamp); err != nil {
+			return Command{action: actionFailed}, fmt.Errorf("context canceled during consolidation TTL")
+		}
+
+		ok, err := c.validateConsolidation(ctx, cmd)
+		if err != nil {
+			return Command{action: actionFailed}, fmt.Errorf("determining candidate node, %w", err)
+		}
+		if !ok {
+			continue
+		}
+		return cmd, nil
+	}
+	// If none of the candidates were able to be deprovisioned, return nothingToDo
+	return Command{action: actionDoNothing}, nil
 }
 
 // deleteEmpty returns a deprovisioningCommmand if there are empty nodes.
-func (c *Consolidation) deleteEmpty(_ context.Context, candidates ...CandidateNode) Command {
+func (c *Consolidation) deleteEmpty(ctx context.Context, validationTimestamp time.Time, candidates ...CandidateNode) (Command, time.Time, error) {
 	emptyNodes := lo.Filter(candidates, func(n CandidateNode, _ int) bool { return len(n.pods) == 0 })
-	if len(emptyNodes) != 0 {
-		return Command{
-			nodesToRemove: lo.Map(emptyNodes, func(n CandidateNode, _ int) *v1.Node { return n.Node }),
-			action:        actionDelete,
-			createdAt:     c.clock.Now(),
-		}
+	if len(emptyNodes) == 0 {
+		return Command{action: actionDoNothing}, validationTimestamp, nil
+	}
+
+	cmd := Command{
+		nodesToRemove: lo.Map(emptyNodes, func(n CandidateNode, _ int) *v1.Node { return n.Node }),
+		action:        actionDelete,
 	}
-	return Command{action: actionDoNothing}
+
+	validationTimestamp = lo.Ternary(validationTimestamp.IsZero(), c.clock.Now(), validationTimestamp)
```

Comment:
You could potentially see if `lo.Coalesce()` works for this, seems interesting 🤔

Reply:
no longer using this line, but that's a cool function!